In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('E:/Datasets/target_action_prediction_SberAvtopodpiska/ga_hits.csv')
df.head(10)

session_id    hit_date   hit_time  \
0  5639623078712724064.1640254056.1640254056  2021-12-23   597864.0   
1  7750352294969115059.1640271109.1640271109  2021-12-23   597331.0   
2   885342191847998240.1640235807.1640235807  2021-12-23   796252.0   
3   142526202120934167.1640211014.1640211014  2021-12-23   934292.0   
4  3450086108837475701.1640265078.1640265078  2021-12-23   768741.0   
5  6466333295973247896.1640246168.1640246168  2021-12-23   704194.0   
6   281215510786615563.1640245516.1640245516  2021-12-23   911854.0   
7  4024492994895054107.1640269084.1640269084  2021-12-23  1071804.0   
8   555009234841130092.1640256620.1640256620  2021-12-23  1013899.0   
9  2692901778487480807.1640206845.1640206845  2021-12-23        0.0   

   hit_number hit_type           hit_referer  \
0          30    event                   NaN   
1          41    event                   NaN   
2          49    event                   NaN   
3          46    event                   NaN   
4          79    event                   NaN   
5          41    event                   NaN   
6         102    event                   NaN   
7          85    event                   NaN   
8         101    event  VloVXNWduHeTjUoDkjkO   
9           1    event                   NaN   

                                       hit_page_path event_category  \
0  sberauto.com/cars?utm_source_initial=google&ut...           quiz   
1  sberauto.com/cars/fiat?city=1&city=18&rental_c...           quiz   
2  sberauto.com/cars/all/volkswagen/polo/e994838f...           quiz   
3  sberauto.com/cars?utm_source_initial=yandex&ut...           quiz   
4  sberauto.com/cars/all/mercedes-benz/cla-klasse...           quiz   
5  sberauto.com/cars?utm_source_initial=youtube&u...           quiz   
6  sberauto.com/cars/toyota?isnew=false&rental_ca...           quiz   
7  sberauto.com/cars/all/mercedes-benz/glc/f8f330...           quiz   
8  sberauto.com/cars/all/kia/sorento/c38179cb?utm...           quiz   
9  sberauto.com/cars/all/nissan/x-trail/0744675f?...       card_web   

  event_action event_label  event_value  
0    quiz_show         NaN          NaN  
1    quiz_show         NaN          NaN  
2    quiz_show         NaN          NaN  
3    quiz_show         NaN          NaN  
4    quiz_show         NaN          NaN  
5    quiz_show         NaN          NaN  
6    quiz_show         NaN          NaN  
7    quiz_show         NaN          NaN  
8    quiz_show         NaN          NaN  
9    view_card         NaN          NaN

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15726470 entries, 0 to 15726469
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   session_id      object 
 1   hit_date        object 
 2   hit_time        float64
 3   hit_number      int64  
 4   hit_type        object 
 5   hit_referer     object 
 6   hit_page_path   object 
 7   event_category  object 
 8   event_action    object 
 9   event_label     object 
 10  event_value     float64
dtypes: float64(2), int64(1), object(8)
memory usage: 1.3+ GB


In [4]:
df.describe(include='all')

session_id    hit_date      hit_time  \
count                                    15726470    15726470  6.566148e+06   
unique                                    1734610         226           NaN   
top     5442565791571325612.1632449195.1632449195  2021-05-24           NaN   
freq                                          768      513035           NaN   
mean                                          NaN         NaN  2.091050e+05   
std                                           NaN         NaN  4.032110e+05   
min                                           NaN         NaN  0.000000e+00   
25%                                           NaN         NaN  2.412000e+04   
50%                                           NaN         NaN  8.994100e+04   
75%                                           NaN         NaN  2.195352e+05   
max                                           NaN         NaN  1.315688e+07   

          hit_number  hit_type           hit_referer           hit_page_path  \
count   1.572647e+07  15726470               9451666                15726470   
unique           NaN         1                 37873                  342715   
top              NaN     event  HbolMJUevblAbkHClEQa  podpiska.sberauto.com/   
freq             NaN  15726470               8879187                 2793639   
mean    2.356715e+01       NaN                   NaN                     NaN   
std     2.887713e+01       NaN                   NaN                     NaN   
min     1.000000e+00       NaN                   NaN                     NaN   
25%     7.000000e+00       NaN                   NaN                     NaN   
50%     1.500000e+01       NaN                   NaN                     NaN   
75%     2.900000e+01       NaN                   NaN                     NaN   
max     5.000000e+02       NaN                   NaN                     NaN   

       event_category event_action           event_label  event_value  
count        15726470     15726470              11966286          0.0  
unique             52          230                 39825          NaN  
top          card_web    view_card  KclpemfoHstknWHFiLit          NaN  
freq          7456998      3558985               6505447          NaN  
mean              NaN          NaN                   NaN          NaN  
std               NaN          NaN                   NaN          NaN  
min               NaN          NaN                   NaN          NaN  
25%               NaN          NaN                   NaN          NaN  
50%               NaN          NaN                   NaN          NaN  
75%               NaN          NaN                   NaN          NaN  
max               NaN          NaN                   NaN          NaN

In [5]:
df2 = pd.read_csv('E:/Datasets/target_action_prediction_SberAvtopodpiska/ga_sessions.csv', low_memory=False)
df2.head(10)

session_id              client_id  \
0  9055434745589932991.1637753792.1637753792  2108382700.1637753791   
1   905544597018549464.1636867290.1636867290   210838531.1636867288   
2  9055446045651783499.1640648526.1640648526  2108385331.1640648523   
3  9055447046360770272.1622255328.1622255328  2108385564.1622255328   
4  9055447046360770272.1622255345.1622255345  2108385564.1622255328   
5  9055447192389856083.1622453074.1622453074  2108385598.1622453075   
6  9055455318486370642.1640843788.1640843788  2108387490.1640843602   
7  9055461992850812764.1626107740.1626107740  2108389044.1626107740   
8  9055462349345527315.1638536723.1638536723  2108389127.1638536723   
9  9055466554104774132.1624800757.1624800757  2108390106.1624800756   

   visit_date visit_time  visit_number            utm_source utm_medium  \
0  2021-11-24   14:36:32             1  ZpYIoDJMcFzVoPFsHGJL     banner   
1  2021-11-14   08:21:30             1  MvfHsxITijuriZxsqZqt        cpm   
2  2021-12-28   02:42:06             1  ZpYIoDJMcFzVoPFsHGJL     banner   
3  2021-05-29   05:00:00             1  kjsLglQLzykiRbcDiGcD        cpc   
4  2021-05-29   05:00:00             2  kjsLglQLzykiRbcDiGcD        cpc   
5  2021-05-31   12:00:00             1  kjsLglQLzykiRbcDiGcD    organic   
6  2021-12-30   08:56:28             1  TxKUcPpthBDPieTGmVhx        cpc   
7  2021-07-12   19:00:00             1  ZpYIoDJMcFzVoPFsHGJL     banner   
8  2021-12-03   16:05:23             1  ZpYIoDJMcFzVoPFsHGJL     banner   
9  2021-06-27   16:00:00             1  ZpYIoDJMcFzVoPFsHGJL     banner   

           utm_campaign         utm_adcontent           utm_keyword  \
0  LEoPHuyFvzoNfnzGgfcd  vCIpmpaGBnIQhyYNkXqp  puhZPIYqKXeFPaUviSjo   
1  FTjNLDyTrXaWYgZymFkV  xhoenQgDQsgfEPYNPwKO  IGUCNvHlhfHpROGclCit   
2  LEoPHuyFvzoNfnzGgfcd  vCIpmpaGBnIQhyYNkXqp  puhZPIYqKXeFPaUviSjo   
3                   NaN  NOBKLgtuvqYWkXQHeYWM                   NaN   
4                   NaN                   NaN                   NaN   
5  LTuZkdKfxRGVceoWkVyg  JNHcPlZPxEMWDnRiyoBf                   NaN   
6  FTjNLDyTrXaWYgZymFkV  LcGIUNPUAmXtQJaDfFBR  PwscUHjoUJDrtfWESIHj   
7  LEoPHuyFvzoNfnzGgfcd  JNHcPlZPxEMWDnRiyoBf                   NaN   
8  LEoPHuyFvzoNfnzGgfcd  vCIpmpaGBnIQhyYNkXqp  puhZPIYqKXeFPaUviSjo   
9  LEoPHuyFvzoNfnzGgfcd  JNHcPlZPxEMWDnRiyoBf                   NaN   

  device_category device_os device_brand device_model  \
0          mobile   Android       Huawei          NaN   
1          mobile   Android      Samsung          NaN   
2          mobile   Android       Huawei          NaN   
3          mobile       NaN       Xiaomi          NaN   
4          mobile       NaN       Xiaomi          NaN   
5          mobile       NaN        Apple          NaN   
6          tablet   Android       Lenovo          NaN   
7          mobile       NaN      Samsung          NaN   
8          mobile       iOS        Apple          NaN   
9          mobile       NaN      Samsung          NaN   

  device_screen_resolution    device_browser geo_country          geo_city  
0                  360x720            Chrome      Russia          Zlatoust  
1                  385x854  Samsung Internet      Russia            Moscow  
2                  360x720            Chrome      Russia       Krasnoyarsk  
3                  393x786            Chrome      Russia            Moscow  
4                  393x786            Chrome      Russia            Moscow  
5                  375x812            Safari      Russia  Saint Petersburg  
6                 602x1029         YaBrowser      Russia  Saint Petersburg  
7                  360x640            Chrome      Russia  Saint Petersburg  
8                  390x844            Safari      Russia            Moscow  
9                  412x915            Chrome      Russia            Moscow

In [6]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1860042 entries, 0 to 1860041
Data columns (total 18 columns):
 #   Column                    Dtype 
---  ------                    ----- 
 0   session_id                object
 1   client_id                 object
 2   visit_date                object
 3   visit_time                object
 4   visit_number              int64 
 5   utm_source                object
 6   utm_medium                object
 7   utm_campaign              object
 8   utm_adcontent             object
 9   utm_keyword               object
 10  device_category           object
 11  device_os                 object
 12  device_brand              object
 13  device_model              object
 14  device_screen_resolution  object
 15  device_browser            object
 16  geo_country               object
 17  geo_city                  object
dtypes: int64(1), object(17)
memory usage: 255.4+ MB


In [7]:
df2.describe(include='all')

session_id              client_id  \
count                                     1860042                1860042   
unique                                    1860042                1391719   
top     9055434745589932991.1637753792.1637753792  1750498477.1629450062   
freq                                            1                    462   
mean                                          NaN                    NaN   
std                                           NaN                    NaN   
min                                           NaN                    NaN   
25%                                           NaN                    NaN   
50%                                           NaN                    NaN   
75%                                           NaN                    NaN   
max                                           NaN                    NaN   

        visit_date visit_time  visit_number            utm_source utm_medium  \
count      1860042    1860042  1.860042e+06               1859945    1860042   
unique         226      85318           NaN                   293         56   
top     2021-05-24   12:00:00           NaN  ZpYIoDJMcFzVoPFsHGJL     banner   
freq         39453      61067           NaN                578290     552272   
mean           NaN        NaN  2.712804e+00                   NaN        NaN   
std            NaN        NaN  1.182907e+01                   NaN        NaN   
min            NaN        NaN  1.000000e+00                   NaN        NaN   
25%            NaN        NaN  1.000000e+00                   NaN        NaN   
50%            NaN        NaN  1.000000e+00                   NaN        NaN   
75%            NaN        NaN  2.000000e+00                   NaN        NaN   
max            NaN        NaN  5.640000e+02                   NaN        NaN   

                utm_campaign         utm_adcontent           utm_keyword  \
count                1640439               1524427                777981   
unique                   412                   286                  1219   
top     LTuZkdKfxRGVceoWkVyg  JNHcPlZPxEMWDnRiyoBf  puhZPIYqKXeFPaUviSjo   
freq                  463481               1006599                506819   
mean                     NaN                   NaN                   NaN   
std                      NaN                   NaN                   NaN   
min                      NaN                   NaN                   NaN   
25%                      NaN                   NaN                   NaN   
50%                      NaN                   NaN                   NaN   
75%                      NaN                   NaN                   NaN   
max                      NaN                   NaN                   NaN   

       device_category device_os device_brand          device_model  \
count          1860042    789904      1492864                 16338   
unique               3        13          205                   104   
top             mobile   Android        Apple  AuMdmADEIoPXiWpTsBEj   
freq           1474871    464054       551088                  9778   
mean               NaN       NaN          NaN                   NaN   
std                NaN       NaN          NaN                   NaN   
min                NaN       NaN          NaN                   NaN   
25%                NaN       NaN          NaN                   NaN   
50%                NaN       NaN          NaN                   NaN   
75%                NaN       NaN          NaN                   NaN   
max                NaN       NaN          NaN                   NaN   

       device_screen_resolution device_browser geo_country geo_city  
count                   1860042        1860042     1860042  1860042  
unique                     5039             57         166     2548  
top                     414x896         Chrome      Russia   Moscow  
freq                     169090        1013436     1800565   805329  
mean                        NaN            NaN   

In [8]:
df.duplicated().any()

False

In [9]:
df2.duplicated().any()

False

In [10]:
df.isna().any()

session_id        False
hit_date          False
hit_time           True
hit_number        False
hit_type          False
hit_referer        True
hit_page_path     False
event_category    False
event_action      False
event_label        True
event_value        True
dtype: bool

In [11]:
df2.isna().any()

session_id                  False
client_id                   False
visit_date                  False
visit_time                  False
visit_number                False
utm_source                   True
utm_medium                  False
utm_campaign                 True
utm_adcontent                True
utm_keyword                  True
device_category             False
device_os                    True
device_brand                 True
device_model                 True
device_screen_resolution    False
device_browser              False
geo_country                 False
geo_city                    False
dtype: bool

# 1. EDA.

1.1. Датасет представлен двумя таблицами. В обеих таблицах отсутствуют дубли.

1.2. Исходя из количества уникальных строк сессий в обеих таблицах и общего количества строк хитов становится очевидно, что существуют сессии, как без единого хита, так и сессии где хитов 1 и более. Однако поскольку по условиям задания модель принимает только аттрибуты вида utm_*, device_*, geo_*, то фактически нам из таблицы хитов нужны только значения атрибута event_action, который содержит целевую переменную, а также session_id по которому объединяются таблицы.

1.3. Атрибуты hit_date, hit_time, hit_number, hit_type, hit_referer, hit_page_path, event_category, event_label, event_value не являются аттрибутами модели, поэтому фичи на удаление.

1.4. Идентификатор сессии session_id нужен для объединения таблиц, поэтому после объединения фича идет на удаление.

1.5. client_id, visit_date, visit_time, visit_number не являются аттрибутами модели, поэтому фичи на удаление.

1.6. Информация о канале привлечения utm_source содержит 97 пропусков. При 293 уникальных значениях имеет смысл замены на наиболее часто встречающийся.

1.7. В типе привлечения utm_medium отсутствуют пропуски.

1.8. Идентификатор рекламной компании utm_campaign содержит 219603 пропуска, что составляет около 12% от общего количества. Удаление этой фичи не видится целесообразным, однако например для значения  utm_source = BHcvLfOaCWvWTykYqHVe, что составляет около половины всех пропусков какие-либо данные отсутствуют, поэтому в рекламной компании все пропуски будем заменять на значение other.

1.9. Атрибут utm_adcontent содержит 335615 пропусков, что составляет около 18% от общего количества. Удаление этой фичи не видится целесообразным, однако например для значения  utm_source = BHcvLfOaCWvWTykYqHVe, что составляет около трети всех пропусков какие-либо данные отсутствуют, поэтому в utm_adcontent все пропуски будем заменять на значение other.

1.10. Атрибут utm_keyword содержит пропусков более 50% от общего количества. Фича на удаление.

1.11. Атрибут device_category не содержит пропусков.

1.12. Атрибут device_os содержит пропусков более 50% от общего количества. Фича на удаление.

1.13. Атрибут device_brand содержит 367178 пропусков, что составляет около 20% от общего количества. Удаление этой фичи не видится целесообразным, поэтому в device_brand все пропуски будем заменять на значение other.

1.14.Атрибут device_model подавляющее количество пропусков. Фича на удаление.

1.15. Атрибуты device_screen_resolution, device_browser, geo_country, geo_city не содержат пропусков.